File: bincodes/verify_code.py
#!/usr/bin/env python3
"""Independent verifier for a binary constant-length code.
Usage: verify_code.py <file> <expected_count> <length> <min_distance>
Checks: length of every string, binary alphabet, distinctness, exact count,
and the Hamming distance of EVERY pair (computed character by character)."""
import sys
from itertools import combinations

path, want_count, want_len, want_d = sys.argv[1], int(sys.argv[2]), int(sys.argv[3]), int(sys.argv[4])
lines = [ln.strip() for ln in open(path) if ln.strip() != ""]

problems = []
# 1. lengths
bad_len = [s for s in lines if len(s) != want_len]
if bad_len: problems.append(f"{len(bad_len)} string(s) not length {want_len}: {bad_len[:3]}")
# 2. alphabet
bad_sym = [s for s in lines if set(s) - {"0", "1"}]
if bad_sym: problems.append(f"{len(bad_sym)} string(s) contain non-binary symbols: {bad_sym[:3]}")
# 3. distinctness
if len(set(lines)) != len(lines):
    dup = sorted({s for s in lines if lines.count(s) > 1})
    problems.append(f"duplicate string(s): {dup[:5]}")
# 4. count
if len(lines) != want_count:
    problems.append(f"expected exactly {want_count} strings, found {len(lines)}")
# 5. pairwise distances, char by char
pairs = 0; mind = None; worst = None
hist = {}
for a, b in combinations(lines, 2):
    if len(a) != len(b): continue
    d = sum(1 for x, y in zip(a, b) if x != y)
    pairs += 1
    hist[d] = hist.get(d, 0) + 1
    if mind is None or d < mind: mind, worst = d, (a, b)
if mind is not None and mind < want_d:
    problems.append(f"minimum pairwise distance {mind} < {want_d}, e.g. {worst}")

print(f"file                  : {path}")
print(f"strings found         : {len(lines)} (required exactly {want_count})")
print(f"all length {want_len}          : {not bad_len}")
print(f"all symbols in {{0,1}}   : {not bad_sym}")
print(f"all distinct          : {len(set(lines)) == len(lines)}")
print(f"pairs checked         : {pairs} (expected {want_count*(want_count-1)//2})")
print(f"minimum distance      : {mind} (required >= {want_d})")
print(f"distance distribution : {dict(sorted(hist.items()))}")
for p in problems: print("FAIL:", p)
print("RESULT                :", "PASS" if not problems else "FAIL")
sys.exit(0 if not problems else 1)

File: bincodes/verify_code.c
/* Second, independently written verifier (C).
   usage: verify_code <file> <count> <len> <mindist> */
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
int main(int argc, char **argv){
  FILE *f = fopen(argv[1], "r");
  int want = atoi(argv[2]), L = atoi(argv[3]), D = atoi(argv[4]);
  char buf[512], w[4096][512]; int n = 0, bad = 0;
  while (fscanf(f, "%500s", buf) == 1) { strcpy(w[n++], buf); }
  fclose(f);
  for (int i = 0; i < n; i++) {
    if ((int)strlen(w[i]) != L) { printf("FAIL: string %d has length %zu\n", i, strlen(w[i])); bad = 1; }
    for (int k = 0; w[i][k]; k++)
      if (w[i][k] != '0' && w[i][k] != '1') { printf("FAIL: non-binary symbol in string %d\n", i); bad = 1; break; }
  }
  for (int i = 0; i < n; i++) for (int j = i+1; j < n; j++)
    if (!strcmp(w[i], w[j])) { printf("FAIL: strings %d and %d are equal\n", i, j); bad = 1; }
  if (n != want) { printf("FAIL: found %d strings, wanted exactly %d\n", n, want); bad = 1; }
  int mind = 1 << 30, pairs = 0;
  for (int i = 0; i < n; i++) for (int j = i+1; j < n; j++) {
    int d = 0; for (int k = 0; k < L; k++) if (w[i][k] != w[j][k]) d++;
    pairs++; if (d < mind) mind = d;
  }
  if (mind < D) { printf("FAIL: minimum distance %d < %d\n", mind, D); bad = 1; }
  printf("strings=%d  length=%d  pairs=%d  min distance=%d\n", n, L, pairs, mind);
  printf("RESULT: %s\n", bad ? "FAIL" : "PASS");
  return bad;
}
